src/bidiMapper/domains/context/contextHelper.ts

```diff
@@ -0,0 +1,216 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Protocol } from 'devtools-protocol';
+import { CdpClient } from '../../../cdp';
+import { CommonDataTypes, Log, Script } from '../../bidiProtocolTypes';
+
+// TODO sadym: `dummyContextObject` needed for the running context.
+// Use the proper `executionContextId` instead:
+// https://github.com/GoogleChromeLabs/chromium-bidi/issues/52
+export async function getDummyContextId(cdpClient: CdpClient): Promise<string> {
+  const dummyContextObject = await cdpClient.Runtime.evaluate({
+    expression: '(()=>{return {}})()',
+  });
+  return dummyContextObject.result.objectId!;
+}
+
+/**
+ * Serializes a given CDP object into BiDi, keeping references in the
+ * target's `globalThis`.
+ * @param cdpObject CDP remote object to be serialized.
+ */
+export async function serializeCdpObject(
+  cdpClient: CdpClient,
+  EVALUATOR_SCRIPT: string,
+  cdpObject: Protocol.Runtime.RemoteObject
+): Promise<CommonDataTypes.RemoteValue> {
+  const response = await cdpClient.Runtime.callFunctionOn({
+    functionDeclaration: `${EVALUATOR_SCRIPT}.serialize`,
+    objectId: await getDummyContextId(cdpClient),
+    arguments: [cdpObject],
+    returnByValue: true,
+  });
+
+  if (response.exceptionDetails)
+    // Serialization failed unexpectidely.
+    throw new Error(
+      'Cannot serialize object: ' + response.exceptionDetails.text
+    );
+
+  return response.result.value;
+}
+
+const specifiers = ['%s', '%d', '%i', '%f', '%o', '%O', '%c'];
+
+export function getLogLevel(consoleAPIType: string): Log.LogLevel {
+  if (consoleAPIType in ['error', 'assert']) {
+    return 'error';
+  }
+  if (consoleAPIType in ['debug', 'trace']) {
+    return 'debug';
+  }
+  if (consoleAPIType in ['warning', 'warn']) {
+    return 'warning';
+  }
+  return 'info';
+}
+
+function isFormmatSpecifier(str: string): boolean {
+  return specifiers.some((spec) => str.includes(spec));
+}
+
+/**
+ * @param args input remote values to be format printed
+ * @returns parsed text of the remote values in specific format
+ */
+export function logMessageFormatter(
+  args: CommonDataTypes.RemoteValue[]
+): string {
+  let output = '';
+  const argFormat = (
+    args[0] as { type: string; value: string }
+  ).value.toString();
+  const argValues = args.slice(1, undefined);
+  const tokens = argFormat.split(
+    new RegExp(specifiers.map((spec) => '(' + spec + ')').join('|'), 'g')
+  );
+
+  tokens.map((token) => {
```

Comment:
This looks like it should be a for-of loop instead of a map() call.

Reply:
Updated map method with for-of loop. Thanks.